### PR TITLE
Automate release on merge in develop

### PR DESCRIPTION
**Pull request by mukulmishra18**
We need to automate the release.

This PR automate the release on every PR merge in `develop` branch.

### CODE REVIEW

File: .github/workflows/merge.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+name: Trigger release on merge
```
Comment (offnertho) [inline .github/workflows/merge.yml]: pls also rename the workflow file.

File: .github/workflows/merge.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+run-name: Starting release for ${{ github.actor }} PR merge
```
Comment (offnertho) [inline .github/workflows/merge.yml]: what else could we use? but yeah should also be fine.

File: .github/workflows/merge.yml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+jobs:
+  on_pr_merge:
+    if: github.event.pull_request.merged == true
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout develop
+        uses: actions/checkout@v4
+        with:
+          ref: develop
```
Comment (offnertho) [inline .github/workflows/merge.yml]: we don't need that because the workflow is only triggered on develop.

File: .github/workflows/merge.yml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Read package.json version
+        uses: actions/github-script@v6
+        id: define-package-json-version
+        with:
+          script: |
+            const { version } = require('./package.json')
+            core.info(`performing a ${{ inputs.release-type }} release for existing version ${version}`)
```
Comment (offnertho) [inline .github/workflows/merge.yml]: I guess we can remove that input right?

File: .github/workflows/merge.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            core.setOutput('packageJsonVersion', version)
+
+      - name: Define release version
+        uses: actions/github-script@v6
+        id: define-release-version
+        with:
+          script: |
+            const { defineReleaseVersion } = require('./.github/scripts/defineVersion.js')
+            return defineReleaseVersion({core}, 'minor', undefined, "${{ steps.define-package-json-version.outputs.packageJsonVersion }}" )
```
Comment (offnertho) [inline .github/workflows/merge.yml]: you can also remove the third parameter from the function, so you don't have to pass undefined.

File: .github/workflows/merge.yml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+      - name: Bump package.json version
+        run: |
+          git config --global user.name 'Automated Release'
+          git config --global user.email 'release-automation@bitmovin.com'
+          npm version "${{ fromJson(steps.define-release-version.outputs.result).full }}"
+
+      - name: Update Changelog
```
Comment (offnertho) [inline .github/workflows/merge.yml]: does this step also add a new `development` section?

**Review by offnertho — CHANGES_REQUESTED** [review]
Looks good in general. 

Some small remarks.

File: .github/workflows/merge.yml (hunk 6 above)
Comment (mukulmishra18) [reply .github/workflows/merge.yml]: Nope. Do we need to add that? I think it should not be required?

File: .github/workflows/merge.yml (hunk 2 above)
Comment (mukulmishra18) [reply .github/workflows/merge.yml]: I mean we can add the commit hash, but that might make the message too long 😄

File: .github/scripts/defineVersion.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
+const semver = require('semver');
+
+function getPlayerUiVersion(versionInput) {
+  let targetVersion = versionInput ?? process.env.CI_BRANCH;
+
+  if (!targetVersion) {
+    console.log('no version provided using CI_BRANCH');
+    process.exit(1);
+  }
```
Comment (offnertho) [inline .github/scripts/defineVersion.js]: I think we can remove the CI_BRANCH fallback.

File: .github/workflows/release-on-merge.yml [hunk 8]
```diff
@@ -0,0 +1,64 @@
+name: Trigger release on merge
```
Comment (offnertho) [inline .github/workflows/release-on-merge.yml]: Maybe this should be called `tag-release-version`?

File: .github/workflows/release-on-merge.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+run-name: Starting release for ${{ github.actor }} PR merge
+on:
+  pull_request:
+    types:
+      - closed
+    branches:
+      - develop
+
+jobs:
+  on_pr_merge:
```
Comment (offnertho) [inline .github/workflows/release-on-merge.yml]: I would rather name it what it does, then when it is executed. 

So I guess `trigger-ui-release` or `tag-ui-release` or something would fit better.

File: .github/workflows/release-on-merge.yml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            const stableVersion = '${{ fromJson(steps.define-release-version.outputs.result).full }}'.split('-')[0]
+            const releaseDate = new Date().toISOString().split('T')[0]
+
+            const data = fs.readFileSync('./CHANGELOG.md',{encoding:'utf8', flag:'r'});
+
+            core.info(`Updating ${stableVersion} with date ${releaseDate} in Changelog`);
+
+            const changelogFileContents = updateChangeLog(data, stableVersion, releaseDate);
+
+            fs.writeFileSync('./CHANGELOG.md', changelogFileContents, 'utf-8');
```
Comment (offnertho) [inline .github/workflows/release-on-merge.yml]: what happens if there is no changelog entry? should we fail the release?

File: .github/workflows/merge.yml (hunk 2 above)
Comment (offnertho) [reply .github/workflows/merge.yml]: Yeah makes sense. Lets tick with that.

File: .github/workflows/merge.yml (hunk 6 above)
Comment (offnertho) [reply .github/workflows/merge.yml]: ok

File: .github/workflows/release-on-merge.yml (hunk 10 above)
Comment (mukulmishra18) [reply .github/workflows/release-on-merge.yml]: Yeah, right now if there is no changelog the release will fail.

I think that should be okay, as every PR should have a changelog IMO.

File: .github/workflows/release-on-merge.yml (hunk 9 above)
Comment (mukulmishra18) [reply .github/workflows/release-on-merge.yml]: `trigger-ui-release` make sense 👍

File: .github/workflows/release-on-merge.yml (hunk 10 above)
Comment (mukulmishra18) [reply .github/workflows/release-on-merge.yml]: This step actually pass but later on when we commit the changes it fails as there is nothing to commit.

**Review by offnertho — APPROVED** [review]
LGTM. 

I guess when this is merged we need to cancel the release that is triggered with this?

File: .github/workflows/release-on-merge.yml (hunk 10 above)
Comment (offnertho) [reply .github/workflows/release-on-merge.yml]: 👍

**Pull request merged**